### PR TITLE
DAOS-12031 client: checking access mode for root object open

### PR DESCRIPTION
**Pull request by wangdi1**
Checking access mode for root object open in dfs_global2local().

Checking obj open mode for modification operation.

A few fixes for dfs open mode.

Required-githooks: true

Co-authored-by: Mohamad Chaarawi <mohamad.chaarawi@intel.com>
Signed-off-by: Di Wang <di.wang@intel.com>

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficent testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/client/dfs/dfs.c [hunk 1]
```diff
@@ -2830,7 +2830,7 @@ dfs_global2local(daos_handle_t poh, daos_handle_t coh, int flags, d_iov_t glob,
 	strcpy(dfs->root.name, "/");
 	dfs->root.mode = S_IFDIR | 0755;
 
-	obj_mode = get_daos_obj_mode(flags);
+	obj_mode = get_daos_obj_mode(flags ?: dfs_params->amode);
```
Comment (mchaarawi) [inline src/client/dfs/dfs.c:2833]: i believe this is a C extension? for portability and clarity i would prefer you specify the second operand explicitly.

**Review by mchaarawi — APPROVED** [review]

File: src/client/dfs/dfs.c (hunk 1 above)
Comment (wangdi1) [reply src/client/dfs/dfs.c]: oh, sure.

**Review by mchaarawi — APPROVED** [review]

**Review by jolivier23 — DISMISSED** [review]

**Comment by wangdi1** [comment]
wait for mpifileutil to be fixed, then I will refresh this patch.

**Comment by wangdi1** [comment]
wait for https://github.com/daos-stack/mpifileutils-pkg/pull/27. to be landed.

**Comment by wangdi1** [comment]
daos_vol/h5_suite.py. are failed with timeout, which is known to happen with tcp.

**Review by jolivier23 — APPROVED** [review]

**Comment by wangdi1** [comment]
https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-10832/22/pipeline.   Note here are the test result with "daily_test" flag.

**Review by jolivier23 — APPROVED** [review]

**Review by ryon-jensen — APPROVED** [review]

**Pull request merged**